File: ABC177/C_Sum_of_product_of_pairs/submission2.java
// AC

import java.util.Scanner;

class Main {
  public static void main(String[] args) {

    var m = 1000000000 + 7;

    var sc = new Scanner(System.in);

    var n = sc.nextInt();
    var a = new int[n];
    for (int i = 0; i < n; i++) {
      a[i] = sc.nextInt() % m;
    }

    long sum1 = 0;
    long sum2 = 0;
    for (int i = 0; i < n; i++) {
      sum1 += a[i];
      sum2 += (long)a[i] * a[i] % m;
    }
    sum1 = sum1 % m;
    sum2 = sum2 % m;

    long t = (sum1 * sum1 - sum2) % m;
    long answer;
    if (t % 2 == 0) {
      answer = t / 2;
    } else {
      answer = (t + m) / 2;
    }

    System.out.println(answer);
  }
}

File: ARC097/D_Equals/submission2.scala
// 

// 1以上size以下の整数に特化したUnionFind
class UnionFindInt(size: Int) {

  private[this] val arr = new Array[Int](size + 1);

  def addPair(parent: Int, child: Int): Unit = {
    val pr = fetchRoot(parent);
    val cr = fetchRoot(child);
    if (pr != cr) {
      arr(cr) = pr;
    }
  }

  def fetchRoot(elem: Int): Int = {
    @scala.annotation.tailrec
    def sub(elem: Int, pendings: List[(Int, Int)]): Int = {
      val p = arr(elem);
      if (p == 0) {
        pendings.foreach { case (c, p) =>
          if (p != elem) {
            arr(c) = elem;
          }
        }
        elem;
      } else {
        sub(p, (elem, p) :: pendings);
      }
    }
    sub(elem, Nil);
  }

}

object Main extends App {
  val sc = new java.util.Scanner(System.in);
  val n, m = sc.nextInt();
  val ps = Array.fill(n)(sc.nextInt());
  val xys = Array.fill(m)((sc.nextInt(), sc.nextInt()));

  val unionFind = new UnionFindInt(n);
  xys.foreach { case (x, y) =>
    unionFind.addPair(y, x);
  }

  val answer = ps.zipWithIndex.count { case (p, i) =>
    unionFind.fetchRoot(p) == unionFind.fetchRoot(i + 1);
  }
  println(answer);
}


